From 8bdba9383fb28d331ec931c6fb9856f572d7599a Mon Sep 17 00:00:00 2001
From: Loren Huang <b02279@freescale.com>
Date: Mon, 13 Oct 2014 17:14:59 +0800
Subject: [PATCH 0711/1594] MGS-200 gpu:Correct the CMA allocator logic

commit c4c28fc1f5f1b54f4a21476062e62eb13b2a7efe from
git://git.freescale.com/imx/linux-2.6-imx.git

In original logic, the CMA always allocate memory, but report
allocation failure, it will cause serious memory leak.
Correct logic to fix it.

Date: Oct 13, 2014
Signed-off-by: Loren Huang <b02279@freescale.com>
Acked-by: Shawn Guo
(cherry picked from commit 0625c8b5b72c620269e5fd5feee93d51f4536d49)
---
 .../freescale/gc_hal_kernel_allocator_cma.c        |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/freescale/gc_hal_kernel_allocator_cma.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/freescale/gc_hal_kernel_allocator_cma.c
index d4611d0..1d44b59 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/freescale/gc_hal_kernel_allocator_cma.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/freescale/gc_hal_kernel_allocator_cma.c
@@ -97,7 +97,6 @@ _CMAFSLAlloc(
     )
 {
     gceSTATUS status;
-    gctPOINTER addr = gcvNULL;
     gcsCMA_PRIV_PTR priv = (gcsCMA_PRIV_PTR)Allocator->privateData;
 
     struct mdl_cma_priv *mdl_priv=gcvNULL;
@@ -106,13 +105,14 @@ _CMAFSLAlloc(
     gcmkHEADER_ARG("Mdl=%p NumPages=%d", Mdl, NumPages);
 
     gcmkONERROR(gckOS_Allocate(os, sizeof(struct mdl_cma_priv), (gctPOINTER *)&mdl_priv));
+    mdl_priv->kvaddr = gcvNULL;
 
     mdl_priv->kvaddr = dma_alloc_writecombine(gcvNULL,
             NumPages * PAGE_SIZE,
             &mdl_priv->physical,
             GFP_KERNEL | gcdNOWARN);
 
-    if (addr == gcvNULL)
+    if (mdl_priv->kvaddr == gcvNULL)
     {
         gcmkONERROR(gcvSTATUS_OUT_OF_MEMORY);
     }
-- 
1.7.5.4

